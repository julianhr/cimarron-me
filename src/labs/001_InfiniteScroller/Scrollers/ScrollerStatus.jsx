import React from 'react'
import PropTypes from 'prop-types'
import styled from '@emotion/styled'
import { connect } from 'react-redux'

import { colors } from '~/styles/theme'

const ANIMATION_BUFFER = 30

const Root = styled.div`
  display: flex;
  width: 100%;
  padding: 10px 30px 10px;
`

const Span = styled.span`
  padding: 2px 10px;
`

export class ScrollerStatus extends React.PureComponent {
  static propTypes = {
    entryCount: PropTypes.number,
  }

  static defaultProps = {
    entryCount: 0,
  }

  state = {
    flashState: 'reset',
    fetchCount: 0,
  }

  componentDidUpdate(prevProps, prevState) {
    // reset state when a new scroller method is selected
    if (this.props.entryCount === 0 && this.state.fetchCount > 0) {
      this.setState({ fetchCount: 0, flashState: 'reset' })
    // animate if count has increased
    } else if (prevProps.entryCount < this.props.entryCount) {
      this.setState({ fetchCount: this.state.fetchCount + 1 },
        () => this.canAnimate() && this.animateFlash()
      )
    }
  }

  canAnimate() {
    // don't animate on initial fetch, only when the user
    //  triggers one by scrolling
    return this.state.fetchCount >= 2
  }

  animateFlash() {
    this.setState({ flashState: 'stage' })
    setTimeout(() => { this.setState({ flashState: 'transition' }) }, ANIMATION_BUFFER)
  }

  getSpanStyle() {
    const transDurationMs = 65

    switch (this.state.flashState) {
      case 'reset':
        return {}
      case 'stage':
        return {
          backgroundColor: colors.highlight,
          willChange: 'background-color',
        }
      case 'transition':
        return {
          transition: `background-color ${transDurationMs}ms`,
          backgroundColor: 'transparent',
        }
    }
  }

  renderStatus() {
    const { entryCount } = this.props

    if (entryCount > 0) {
      return (
        <div>
          Records: 
          <Span
            css={this.getSpanStyle()} 
          >
            {entryCount}
          </Span>
        </div>
      )
    }
  }

  render() {
    return (
      <Root>
        {this.renderStatus()}
      </Root>
    )
  }

}

const mapStateToProps = ({ entryCount }) => ({ entryCount })

export default connect(mapStateToProps)(ScrollerStatus)
