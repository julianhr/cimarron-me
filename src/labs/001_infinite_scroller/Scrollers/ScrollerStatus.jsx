import React from 'react'
import PropTypes from 'prop-types'
import styled from '@emotion/styled'
import { connect } from 'react-redux'

import { colors } from '~/styles/theme'


const Root = styled.div`
  display: flex;
  width: 100%;
  padding: 10px 30px 10px;
`

const Span = styled.span`
  padding: 2px 10px;
`

class ScrollerStatus extends React.PureComponent {
  static propTypes = {
    entryCount: PropTypes.number,
  }

  state = {
    spanStyle: {},
    fetchCount: 0,
  }

  componentDidUpdate(prevProps, prevState) {
    if (this.props.entryCount === 0 && this.state.fetchCount > 0) {
      this.setState({ fetchCount: 0, spanStyle: {} })
    } else if (prevProps.entryCount < this.props.entryCount) {
      this.setState({ fetchCount: this.state.fetchCount + 1 },
        () => { this.countFlash() }
      )
    }
  }

  countFlash() {
    if (this.state.fetchCount < 2) { return }

    const buffer = 30

    this.setState({ spanStyle: {
      backgroundColor: colors.heightlight,
      willChange: 'background-color',
    }})

    setTimeout(() => { this.setState({ spanStyle: {
      transition: 'background-color 240ms',
      backgroundColor: 'transparent',
    }}) }, buffer)
  }

  renderStatus() {
    const { entryCount } = this.props

    if (entryCount > 0) {
      return (
        <div>
          Records: 
          <Span
            css={this.state.spanStyle}
          >
            {entryCount}
          </Span>
        </div>
      )
    }
  }

  render() {
    return (
      <Root>
        {this.renderStatus()}
      </Root>
    )
  }

}

const mapStateToProps = ({ entryCount }) => ({ entryCount })

export default connect(mapStateToProps)(ScrollerStatus)
