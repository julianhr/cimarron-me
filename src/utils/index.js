const validApiSubDomains = new Set('api.flask api.rails api.node'.split(' '))

export function urlBuilder(basePath, query={}, basePort, api='api.flask') {
  if (!validApiSubDomains.has(api)) { throw new Error('invalid api argument') }

  basePath = basePath.trim()
  const isProduction = process.env.NODE_ENV === 'production'
  const host = isProduction ? `http://${api}.cimarron.me` : 'http://localhost'
  const port = basePort === undefined ? (isProduction ? '80' : '5000') : basePort
  const path = basePath.length > 0 && basePath[0] === '/' ? basePath.slice(1) : basePath
  const urlBase = `${host}:${port}/${path}`
  const urlQuery = Object.entries(query).map(([key, val]) => `${key}=${val}`).join('&')

  return `${urlBase}/?${urlQuery}`
}

export function clamp(value, min, max) {
  return Math.min( Math.max(value, min), max )
}

export function getRangeArray(min, max, isString=false) {
  const range = []

  for (let num = min; num <= max; num++) {
    range.push( isString ? num.toString() : num )
  }

  return range
}
