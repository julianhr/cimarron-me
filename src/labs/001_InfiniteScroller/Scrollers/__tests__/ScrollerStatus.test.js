import React from 'react'
import { create } from 'react-test-renderer'
import emotionSerializer from 'jest-emotion'

import { ScrollerStatus }  from '../ScrollerStatus'
import { sleep } from '~/__tests__/utils'


describe('ScrollerStatus', () => {
  describe('flash animation', () => {
    test('animation steps', async (done) => {
      expect.addSnapshotSerializer(emotionSerializer)

      const { setState } = ScrollerStatus.prototype
      const mockSetState = jest.fn(setState)
      ScrollerStatus.prototype.setState = mockSetState

      const wrapper = create(<ScrollerStatus />)
      wrapper.getInstance().animateFlash()
      await sleep(100)
      expect(mockSetState.mock.calls).toMatchSnapshot()

      // cleanup
      ScrollerStatus.prototype.setState = setState
      done()
    })

    it('flashState defaults to reset', () => {
      const wrapper = create(<ScrollerStatus />)
      expect(wrapper.getInstance().state['flashState']).toBe('reset')
    })

    it('renders styles correctly', () => {
      expect.addSnapshotSerializer(emotionSerializer)
      const wrapper = create(<ScrollerStatus entryCount={5} />)
      const instance = wrapper.getInstance()

      for (let flashState  of ['reset', 'stage', 'transition']) {
        instance.setState({ flashState })
        expect(wrapper).toMatchSnapshot(flashState)
      }
    })
  })

  test('#canAnimate', async (done) => {
    jest.setTimeout(500)

    // promisify #canAnimate so it can be awaited
    const { canAnimate } = ScrollerStatus.prototype
    const promise = new Promise(resolve => {
      ScrollerStatus.prototype.canAnimate = function() {
        resolve(this.state)
        return canAnimate.call(this)
      }
    })
    // mock #animateFlash to test #canAnimate is gatekeeping correctly
    const { animateFlash } = ScrollerStatus.prototype
    const mockAnimateFlash = jest.fn(animateFlash)
    ScrollerStatus.prototype.animateFlash = mockAnimateFlash
    // set up
    const wrapper = create(<ScrollerStatus />)
    const instance = wrapper.getInstance()
    const testCases = [
      [0, false, 0, 0], // initial state
      [5, false, 1, 0], // first fetch
      [10, true, 2, 1], // second fetch
      [15, true, 3, 2], // more than 2 fetches
      [0, false, 0, 2], // new scroller selected
    ]

    for (let [entryCount, shouldAnimate, fetchCount, callCountMock] of testCases) {
      if (instance.props['entryCount'] !== entryCount) {
        // wrapper.setProps({ entryCount })
        wrapper.update(<ScrollerStatus entryCount={entryCount} />)
        await promise
      }
      expect(instance.state['fetchCount']).toEqual(fetchCount)
      expect(mockAnimateFlash).toBeCalledTimes(callCountMock)
      expect(canAnimate.call(instance)).toBe(shouldAnimate)
    }

    // cleanup
    ScrollerStatus.prototype.canAnimate = canAnimate
    ScrollerStatus.prototype.animateFlash = animateFlash
    done()
  })
})
