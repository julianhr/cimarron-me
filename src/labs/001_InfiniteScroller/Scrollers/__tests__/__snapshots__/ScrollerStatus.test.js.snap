// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScrollerStatus flash animation animation steps 1`] = `
Array [
  Array [
    Object {
      "flashState": "stage",
    },
  ],
  Array [
    Object {
      "flashState": "transition",
    },
  ],
]
`;

exports[`ScrollerStatus flash animation renders styles correctly: reset 1`] = `
.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  padding: 10px 30px 10px;
}

.emotion-0 {
  padding: 2px 10px;
}

<div
  className="emotion-2 emotion-3"
>
  <div>
    Records:
    <span
      className="emotion-0 emotion-1"
    >
      5
    </span>
  </div>
</div>
`;

exports[`ScrollerStatus flash animation renders styles correctly: stage 1`] = `
.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  padding: 10px 30px 10px;
}

.emotion-0 {
  padding: 2px 10px;
  background-color: #FFD342;
  will-change: background-color;
}

<div
  className="emotion-2 emotion-3"
>
  <div>
    Records:
    <span
      className="emotion-0 emotion-1"
    >
      5
    </span>
  </div>
</div>
`;

exports[`ScrollerStatus flash animation renders styles correctly: transition 1`] = `
.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  padding: 10px 30px 10px;
}

.emotion-0 {
  padding: 2px 10px;
  -webkit-transition: background-color 65ms;
  transition: background-color 65ms;
  background-color: transparent;
}

<div
  className="emotion-2 emotion-3"
>
  <div>
    Records:
    <span
      className="emotion-0 emotion-1"
    >
      5
    </span>
  </div>
</div>
`;
